### PR TITLE
Switch gitVersion format to non-abbreviated hash

### LINKED ISSUE #99376: gitVersion results in inconsistent string values
Ref. https://github.com/google/ko/issues/315#issuecomment-783718564

[pkg/version/base.go](https://github.com/kubernetes/client-go/blob/12449b5492ee1108d29d9ee17b428c84a6a3f55b/pkg/version/base.go#L58) includes:

	// NOTE: The $Format strings are replaced during 'git archive' thanks to the
	// companion .gitattributes file containing 'export-subst' in this same
	// directory.  See also https://git-scm.com/docs/gitattributes
	gitVersion   string = "v0.0.0-master+$Format:%h$"

This `$Format:%h$` directive, activated by `git archive` and the [`.gitattributes`](https://github.com/kubernetes/client-go/blob/https://github.com/kubernetes/client-go/blob/12449b5492ee1108d29d9ee17b428c84a6a3f55b/pkg/version/.gitattributes/pkg/version/.gitattributes) containing the line `base.go export-subst`, results in file contents served by GitHub Archives that can change in unexpected ways. For example, the length of the commit SHA can expand over time as there are collisions.

This non-determinism can lead to surprises, such as https://github.com/google/ko/issues/315, where a client-go dependency resulted in unexpectedly different tar hashes, which was detected when a new version of the ko binary was being added to Homebrew.

tl;dr: though the client-go dependency itself didn't change for ko, later runs of `git archive` (via GitHub Archives) resulted in different contents for the `gitVersion` string to avoid a new collision for the previous prefix length, which cascaded into different content SHAs and shady-looking binary SHAs.

I'm not sure, but after reading docs I believe you could naively solve this by replacing `$Format:%h$` with `$Format:%H$` so the `gitVersion` string includes the full commit hash instead of the abbreviated one.

I have no idea if anything depends on `gitVersion` being a shorter string, but presumably things are fine with it having been a variable-length string so 🤷‍♂️ . Given enough time/collisions, I suppose theoretically `%h` will slow grow to be equal to `%H`, but I don't think that will happen during any of our lifetimes 👴 

cc @Bo98

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:
Fixes #99376

Ensures that source archives for a given sha/tag remain consistent over time, even if sha collisions require the abbreviated hash to grow in the future.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

Not really sure who to target this at, but maybe @kubernetes/sig-release for review? The artifacts we build don't even make use of the git archive mechanism, and our binaries override the variable using build flags.

### CODE REVIEW

**Comment by BenTheElder** [comment]
Are we going to have surprises when this doesn't match the version in other aspects of the build?

**Review by hasheddan — COMMENTED** [review]
/triage accepted
/sig release

**Comment by liggitt** [comment]
> Are we going to have surprises when this doesn't match the version in other aspects of the build?

it already doesn't... the build pins to a 14-character abbreviated hash:

https://github.com/kubernetes/kubernetes/blob/9d7d3eb820c5297b5fcd5ffa21399c8bf2da744a/hack/lib/version.sh#L68-L69

**Comment by liggitt** [comment]
ha, and the build script bit which *does* honor git archive substitution was already using %H:

https://github.com/kubernetes/kubernetes/blob/9d7d3eb820c5297b5fcd5ffa21399c8bf2da744a/hack/lib/version.sh#L40-L54

**Comment by BenTheElder** [comment]
we're hitting I/O throttling in CI again, xref: https://github.com/kubernetes/test-infra/pull/20966#issuecomment-784523335
/retest

**Review by BenTheElder — APPROVED** [review]
/hold
(for others to review)

**Review by hasheddan — APPROVED** [review]
/lgtm
/cc @justaugustus @saschagrunert

**Comment by liggitt** [comment]
plan to unhold tomorrow if there are no other comments

**Comment by dims** [comment]
/approve
/lgtm

sounds like a plan @liggitt
